### PR TITLE
accesslogs: add CEL-based extension filter

### PR DESCRIPTION
**Pull request by douglas-reid**
Commit Message: add CEL-based extension filter for access logging
Additional Description: This PR establishes the ability to filter access log production via CEL expressions over the set of Envoy attributes. This can simply the creation of Envoy access log filters, allowing complex tailoring.

Risk Level: low
Testing: unit

Signed-off-by: Douglas Reid douglas-reid@users.noreply.github.com

### CODE REVIEW

**Comment by phlax** [comment]
hi @douglas-reid you will need to ensure that all files end with a newline (as opposed to an empty line)

you probably want to set your editor to ensure

- newline endings
- no mixed tabs/spaces
- no trailing whitespace

File: source/extensions/extensions_metadata.yaml
Comment (phlax) [inline source/extensions/extensions_metadata.yaml:8]: there isnt currently a category for this - not sure if this should just be `envoy.access_loggers` or whether there is a case for adding such a category

**Comment by douglas-reid** [comment]
> hi @douglas-reid you will need to ensure that all files end with a newline (as opposed to an empty line)

@phlax thanks for the quick response. i had hoped the clang-format steps were taking care of that stuff. I've updated all the files now I believe.

File: source/extensions/extensions_metadata.yaml
Comment (douglas-reid) [reply source/extensions/extensions_metadata.yaml:8]: I wasn't sure either. But just adding it in `envoy.access_loggers` didn't quite seem right either (what if there is a future access logger that uses CEL expressions for the log format, or similar?). Don't have strong feelings tho, so happy to update as makes sense.

**Comment by phlax** [comment]
>  i had hoped the clang-format steps were taking care of that stuff.

these 3 formatting tests are language-agnostic, so happen before and separately to any clang formatting checks

> I've updated all the files now I believe.

there is still a couple with newline issues - my strong recommendation would be to set your editor to automatically fix these 3 issues on save

File: source/extensions/extensions_metadata.yaml
Comment (phlax) [reply source/extensions/extensions_metadata.yaml:8]: yep, that makes sense to me

extension "categories" in this context are kinda the extension points, so there would be some additional code overhead setting that up (beyond my immediate knowledge on how to do it im afraid)

**Comment by douglas-reid** [comment]
> there is still a couple with newline issues - my strong recommendation would be to set your editor to automatically fix these 3 issues on save

thanks for your patience. I thought I had done that, but must have missed a setting somewhere. hopefully, this should be resolved now (based on local runs of glint)

**Comment by douglas-reid** [comment]
@phlax all the tests are passing now.

**Comment by phlax** [comment]
ill have to pass it on for final review to one of the other maintainers (cc @dio @htuch )

i think you will need to merge main (version history changed with last release)

also im wondering if its necessary to change the `requirements.txt` file - afaict the additional dependencies arent used  (also these dependences are available from `base_pip3` )

**Comment by kyessenov** [comment]
LGTM.

**Review by dio — DISMISSED** [review]

**Comment by dio** [comment]
@zuercher sorry, do you think you have time to take a look at this one? Thank you!

**Comment by douglas-reid** [comment]
Gentle ping here.

File: api/envoy/config/accesslog/v3/accesslog.proto
Comment (zuercher) [inline api/envoy/config/accesslog/v3/accesslog.proto:86]: I think the category is actually `envoy.access_loggers.extension_filters`, per https://github.com/envoyproxy/envoy/blob/main/source/common/access_log/access_log_impl.h#L272

That name seems a little redundant, but I think we need to match it or else we need to undertake some additional complexity to deprecate the old category and introduce a more succinct one. (We had a deprecation facility for extension names, but not sure about the categories.)

File: source/extensions/access_loggers/filters/cel/config.cc
Comment (zuercher) [inline source/extensions/access_loggers/filters/cel/config.cc:40]: I think we should throw an exception here. If I understand correctly, this is a case of someone configuring an extension that requires CEL on a build that doesn't support CEL and it seems like that should be a config-time error.

Further, looking at the code that calls this, there are definitely cases where the result of this call is blindly added to a list of filters (e.g. in the `OperatorFilter` constructor), which seems like it'll probably just cause a crash later (in say, `OrFilter::evaluate`).

File: source/extensions/extensions_metadata.yaml
Comment (zuercher) [reply source/extensions/extensions_metadata.yaml:8]: `envoy.access_loggers.extension_filters` here too

File: tools/extensions/extensions_check.py
Comment (zuercher) [inline tools/extensions/extensions_check.py:58]: Extension category here as well.

**Review by zuercher — COMMENTED** [review]
Sorry I missed the tag earlier. This looks good generally, but I think the extension category name is incorrect and I had a comment about the behavior when CEL is not available in the build.

File: api/envoy/config/accesslog/v3/accesslog.proto
Comment (douglas-reid) [reply api/envoy/config/accesslog/v3/accesslog.proto]: OK. I updated all instances to match.

File: source/extensions/access_loggers/filters/cel/config.cc
Comment (douglas-reid) [reply source/extensions/access_loggers/filters/cel/config.cc]: Updated to throw an exception.

File: tools/extensions/extensions_check.py
Comment (douglas-reid) [reply tools/extensions/extensions_check.py]: done.

**Comment by douglas-reid** [comment]
@zuercher please take another look. i've tried to address your feedback.

File: source/extensions/access_loggers/filters/cel/config.cc
Comment (zuercher) [inline source/extensions/access_loggers/filters/cel/config.cc:21]: Missed this before: Our todo format is `// TODO(github_handle):` so to make it easier to track down the person with context.

**Review by zuercher — COMMENTED** [review]
One tiny thing, but otherwise looks good to me.

File: source/extensions/access_loggers/filters/cel/config.cc
Comment (douglas-reid) [reply source/extensions/access_loggers/filters/cel/config.cc]: updated.

**Comment by douglas-reid** [comment]
@zuercher thanks for the reviews!  if this looks good to you now, i'll rebase and resolve the conflict.

**Comment by zuercher** [comment]
@douglas-reid yeah, go for it

**Comment by douglas-reid** [comment]
ok. rebased and pushed. hopefully 100% ready at this point.

**Review by zuercher — APPROVED** [review]
Thanks.

@markdroth This needs another API bump since we fixed the extension category.
@moderation Could you review the dependency change (I don't think it's really a dependency change other than allowing access for a new extension).

**Comment by zuercher** [comment]
@envoyproxy/api-shepherds I think this is waiting on a re-review.

**Comment by douglas-reid** [comment]
@zuercher It looks like a conflict has appeared again in the intervening time.

What is the right way to get this in?  Should I rebase again and start the wait over, or should I hold off until the final approval is granted again before rebasing?

**Comment by zuercher** [comment]
Don't rebase it. Just merge main and fix the conflict. We'll squash the commit to master. I'll try another means to get api re-approval.

**Comment by zuercher** [comment]
(I don't think that worked, unfortunately.)

**Comment by douglas-reid** [comment]
@zuercher ok. merged and pushed. ptal.

**Review by zuercher — APPROVED** [review]

**Comment by douglas-reid** [comment]
Thanks everyone, for your patience and guidance on this PR. Excited to have the capability merged!

**Pull request merged**
